File: smartedu-backend/src/main/java/com/smartedu/controller/MarksController.java
package com.smartedu.controller;

import com.smartedu.model.Marks;
import com.smartedu.service.MarksService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/api/marks")
public class MarksController {

    @Autowired
    private MarksService marksService;

    @PostMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Marks> addMarks(@RequestBody Marks marks) {
        return ResponseEntity.ok(marksService.addMarks(marks));
    }

    @GetMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<List<Marks>> getAllMarks() {
        return ResponseEntity.ok(marksService.getAllMarks());
    }

    @GetMapping("/student/{studentId}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER') or hasRole('STUDENT')")
    public ResponseEntity<List<Marks>> getMarksByStudent(@PathVariable Long studentId) {
        org.springframework.security.core.Authentication auth = org.springframework.security.core.context.SecurityContextHolder.getContext().getAuthentication();
        boolean isStudent = auth.getAuthorities().stream().anyMatch(a -> a.getAuthority().equals("ROLE_STUDENT"));
        if (isStudent) {
            return ResponseEntity.ok(marksService.getApprovedMarksByStudent(studentId));
        }
        return ResponseEntity.ok(marksService.getMarksByStudent(studentId));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<Marks> updateMarks(@PathVariable Long id, @RequestBody Marks marksDetails) {
        return ResponseEntity.ok(marksService.updateMarks(id, marksDetails));
    }

    @PutMapping("/{id}/approve")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Marks> approveMarks(@PathVariable Long id) {
        return ResponseEntity.ok(marksService.updateApprovalStatus(id, com.smartedu.model.RecordStatus.APPROVED));
    }

    @PutMapping("/{id}/reject")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<Marks> rejectMarks(@PathVariable Long id) {
        return ResponseEntity.ok(marksService.updateApprovalStatus(id, com.smartedu.model.RecordStatus.REJECTED));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN') or hasRole('TEACHER')")
    public ResponseEntity<?> deleteMarks(@PathVariable Long id) {
        marksService.deleteMarks(id);
        return ResponseEntity.ok().build();
    }
}

File: smartedu-backend/controllers/marksController.js
const marksService = require('../services/marksService');

async function addMarks(req, res, next) {
    try {
        const marks = await marksService.addMarks(req.body, Boolean(req.user?.isDemo));
        return res.json(marks);
    } catch (error) {
        next(error);
    }
}

async function getAllMarks(req, res, next) {
    try {
        const marksList = await marksService.getAllMarks(Boolean(req.user?.isDemo));
        return res.json(marksList);
    } catch (error) {
        next(error);
    }
}

async function getMarksByStudent(req, res, next) {
    try {
        const { studentId } = req.params;
        const userRole = req.user.role ? req.user.role.toUpperCase().replace(/^ROLE_/, '') : '';
        const isDemo = Boolean(req.user?.isDemo);
        if (userRole === 'STUDENT') {
            const marksList = await marksService.getApprovedMarksByStudent(studentId, isDemo);
            return res.json(marksList);
        }
        const marksList = await marksService.getMarksByStudent(studentId, isDemo);
        return res.json(marksList);
    } catch (error) {
        next(error);
    }
}

async function updateMarks(req, res, next) {
    try {
        const { id } = req.params;
        const marks = await marksService.updateMarks(id, req.body, Boolean(req.user?.isDemo));
        return res.json(marks);
    } catch (error) {
        next(error);
    }
}

async function approveMarks(req, res, next) {
    try {
        const { id } = req.params;
        const marks = await marksService.updateApprovalStatus(id, 'APPROVED', Boolean(req.user?.isDemo));
        return res.json(marks);
    } catch (error) {
        next(error);
    }
}

async function rejectMarks(req, res, next) {
    try {
        const { id } = req.params;
        const marks = await marksService.updateApprovalStatus(id, 'REJECTED', Boolean(req.user?.isDemo));
        return res.json(marks);
    } catch (error) {
        next(error);
    }
}

async function deleteMarks(req, res, next) {
    try {
        const { id } = req.params;
        await marksService.deleteMarks(id, Boolean(req.user?.isDemo));
        return res.json({ success: true, message: 'Marks record deleted successfully' });
    } catch (error) {
        next(error);
    }
}

module.exports = {
    addMarks,
    getAllMarks,
    getMarksByStudent,
    updateMarks,
    approveMarks,
    rejectMarks,
    deleteMarks
};
